File: backend/app/models/replenishment.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from sqlalchemy import CheckConstraint, JSON, Boolean, Date, DateTime, ForeignKey, Index, Numeric, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.identity import TimestampMixin


QTY = Numeric(18, 6)


class ReplenishmentPolicy(Base, TimestampMixin):
    __tablename__ = "replenishment_policies"
    __table_args__ = (
        CheckConstraint("fixed_target_qty IS NULL OR fixed_target_qty >= 0", name="ck_replenishment_policy_fixed_target_nonnegative"),
        CheckConstraint("min_batch_qty IS NULL OR min_batch_qty > 0", name="ck_replenishment_policy_min_batch_positive"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), unique=True, index=True)
    algorithm: Mapped[str] = mapped_column(String(40), default="SIX_MONTH_MAX", index=True)
    rounding_mode: Mapped[str] = mapped_column(String(30), default="NONE")
    fixed_target_qty: Mapped[Decimal | None] = mapped_column(QTY)
    six_month_weights: Mapped[list[str] | None] = mapped_column(JSON)
    min_batch_qty: Mapped[Decimal | None] = mapped_column(QTY)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    note: Mapped[str | None] = mapped_column(Text)
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    updated_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))


class ReplenishmentRun(Base, TimestampMixin):
    __tablename__ = "replenishment_runs"
    __table_args__ = (
        Index("ix_replenishment_runs_fingerprint_status", "input_fingerprint", "status"),
        CheckConstraint("default_fixed_target_qty IS NULL OR default_fixed_target_qty >= 0", name="ck_replenishment_run_fixed_target_nonnegative"),
        CheckConstraint("default_min_batch_qty IS NULL OR default_min_batch_qty > 0", name="ck_replenishment_run_min_batch_positive"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    run_no: Mapped[str] = mapped_column(String(40), unique=True, index=True)
    calculation_date: Mapped[date] = mapped_column(Date, index=True)
    shipment_batch_id: Mapped[int] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    inventory_batch_id: Mapped[int] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    pipe_wip_batch_id: Mapped[int] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    fitting_wip_batch_id: Mapped[int] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    regular_product_batch_id: Mapped[int] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    weekly_plan_batch_id: Mapped[int | None] = mapped_column(ForeignKey("import_batches.id", ondelete="RESTRICT"), index=True)
    input_fingerprint: Mapped[str] = mapped_column(String(64), index=True)
    status: Mapped[str] = mapped_column(String(30), default="DRAFT", index=True)
    formula_version: Mapped[str] = mapped_column(String(30), default="PHASE3_V1")
    default_algorithm: Mapped[str] = mapped_column(String(40), default="SIX_MONTH_MAX")
    default_weight_config: Mapped[list[str] | None] = mapped_column(JSON)
    default_fixed_target_qty: Mapped[Decimal | None] = mapped_column(QTY)
    rounding_mode: Mapped[str] = mapped_column(String(30), default="NONE")
    default_min_batch_qty: Mapped[Decimal | None] = mapped_column(QTY)
    source_snapshot: Mapped[dict[str, Any]] = mapped_column(JSON)
    source_date_summary: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    calculation_config: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    override_reason: Mapped[str | None] = mapped_column(Text)
    error_summary: Mapped[dict[str, Any] | None] = mapped_column(JSON)
    total_products: Mapped[int] = mapped_column(default=0)
    suggestion_count: Mapped[int] = mapped_column(default=0)
    positive_suggestion_count: Mapped[int] = mapped_column(default=0)
    pending_review_count: Mapped[int] = mapped_column(default=0)
    reviewed_count: Mapped[int] = mapped_column(default=0)
    blocking_issue_count: Mapped[int] = mapped_column(default=0)
    warning_issue_count: Mapped[int] = mapped_column(default=0)
    warning_count: Mapped[int] = mapped_column(default=0)
    approved_count: Mapped[int] = mapped_column(default=0)
    converted_count: Mapped[int] = mapped_column(default=0)
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True)
    calculated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    approved_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    cancelled_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    cancelled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    cancel_reason: Mapped[str | None] = mapped_column(Text)


class ReplenishmentSuggestion(Base, TimestampMixin):
    __tablename__ = "replenishment_suggestions"
    __table_args__ = (
        UniqueConstraint("run_id", "product_id", name="uq_replenishment_suggestion_run_product"),
        CheckConstraint("confirmed_qty IS NULL OR confirmed_qty >= 0", name="ck_replenishment_suggestion_confirmed_nonnegative"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("replenishment_runs.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"), index=True)
    algorithm: Mapped[str] = mapped_column(String(40), index=True)
    algorithm_config: Mapped[dict[str, Any]] = mapped_column(JSON)
    policy_snapshot: Mapped[dict[str, Any]] = mapped_column(JSON)
    monthly_qty_json: Mapped[dict[str, str]] = mapped_column(JSON)
    monthly_shipments: Mapped[dict[str, str]] = mapped_column(JSON)
    six_month_max: Mapped[Decimal] = mapped_column(QTY)
    six_month_avg: Mapped[Decimal] = mapped_column(QTY)
    three_month_avg: Mapped[Decimal] = mapped_column(QTY)
    weighted_avg: Mapped[Decimal] = mapped_column(QTY)
    order_input_qty: Mapped[Decimal] = mapped_column(QTY)
    fixed_target_qty: Mapped[Decimal | None] = mapped_column(QTY)
    calculated_target_qty: Mapped[Decimal] = mapped_column(QTY)
    target_stock_qty: Mapped[Decimal] = mapped_column(QTY)
    on_hand_qty: Mapped[Decimal] = mapped_column(QTY)
    expected_inbound_qty: Mapped[Decimal] = mapped_column(QTY)
    expected_outbound_qty: Mapped[Decimal] = mapped_column(QTY)
    available_qty: Mapped[Decimal] = mapped_column(QTY)
    pipe_wip_raw_qty: Mapped[Decimal] = mapped_column(QTY)
    pipe_wip_effective_qty: Mapped[Decimal] = mapped_column(QTY)
    fitting_wip_raw_qty: Mapped[Decimal] = mapped_column(QTY)
    fitting_wip_effective_qty: Mapped[Decimal] = mapped_column(QTY)
    scheduled_not_started_qty: Mapped[Decimal] = mapped_column(QTY)
    scheduled_known_qty: Mapped[Decimal] = mapped_column(QTY)
    scheduled_override_qty: Mapped[Decimal] = mapped_column(QTY)
    scheduled_source_status: Mapped[str] = mapped_column(String(30), default="NONE")
    raw_suggested_qty: Mapped[Decimal] = mapped_column(QTY)
    system_suggested_qty: Mapped[Decimal] = mapped_column(QTY)
    confirmed_qty: Mapped[Decimal | None] = mapped_column(QTY)
    review_status: Mapped[str] = mapped_column(String(30), default="PENDING", index=True)
    reviewed_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    review_reason: Mapped[str | None] = mapped_column(Text)
    change_reason: Mapped[str | None] = mapped_column(Text)
    converted_demand_id: Mapped[int | None] = mapped_column(
        ForeignKey(
            "production_demands.id", ondelete="SET NULL",
            name="fk_replenishment_suggestions_converted_demand_id", use_alter=True,
        ),
        index=True,
    )


class ReplenishmentIssue(Base, TimestampMixin):
    __tablename__ = "replenishment_issues"
    __table_args__ = (Index("ix_replenishment_issues_run_severity_status", "run_id", "severity", "status"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("replenishment_runs.id", ondelete="CASCADE"), index=True)
    suggestion_id: Mapped[int | None] = mapped_column(ForeignKey("replenishment_suggestions.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int | None] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    issue_code: Mapped[str] = mapped_column(String(100), index=True)
    severity: Mapped[str] = mapped_column(String(20), index=True)
    message: Mapped[str] = mapped_column(Text)
    details: Mapped[dict[str, Any] | None] = mapped_column(JSON)
    status: Mapped[str] = mapped_column(String(20), default="OPEN", index=True)
    resolved_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    resolution_note: Mapped[str | None] = mapped_column(Text)


class ReplenishmentOrderInput(Base, TimestampMixin):
    __tablename__ = "replenishment_order_inputs"
    __table_args__ = (
        UniqueConstraint("run_id", "product_id", name="uq_replenishment_order_run_product"),
        CheckConstraint("order_qty >= 0", name="ck_replenishment_order_qty_nonnegative"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("replenishment_runs.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    order_qty: Mapped[Decimal] = mapped_column(QTY)
    source_document_no: Mapped[str | None] = mapped_column(String(100))
    note: Mapped[str | None] = mapped_column(Text)
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"))


class ProductionDemand(Base, TimestampMixin):
    __tablename__ = "production_demands"
    __table_args__ = (
        CheckConstraint("confirmed_qty > 0", name="ck_production_demand_confirmed_positive"),
        CheckConstraint("active_allocated_qty >= 0", name="ck_production_demand_allocated_nonnegative"),
        CheckConstraint("qualified_completed_qty >= 0", name="ck_production_demand_completed_nonnegative"),
        CheckConstraint("remaining_to_schedule_qty >= 0", name="ck_production_demand_remaining_schedule_nonnegative"),
        CheckConstraint("remaining_to_complete_qty >= 0", name="ck_production_demand_remaining_complete_nonnegative"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    demand_no: Mapped[str] = mapped_column(String(40), unique=True, index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"), index=True)
    source_type: Mapped[str] = mapped_column(String(30), default="REPLENISHMENT")
    source_suggestion_id: Mapped[int] = mapped_column(
        ForeignKey("replenishment_suggestions.id", ondelete="RESTRICT"), unique=True, index=True
    )
    confirmed_qty: Mapped[Decimal] = mapped_column(QTY)
    active_allocated_qty: Mapped[Decimal] = mapped_column(QTY, default=Decimal("0"))
    qualified_completed_qty: Mapped[Decimal] = mapped_column(QTY, default=Decimal("0"))
    remaining_to_schedule_qty: Mapped[Decimal] = mapped_column(QTY)
    remaining_to_complete_qty: Mapped[Decimal] = mapped_column(QTY)
    priority: Mapped[int] = mapped_column(default=0)
    required_date: Mapped[date | None] = mapped_column(Date)
    status: Mapped[str] = mapped_column(String(30), default="PENDING_SCHEDULE", index=True)
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"))
    cancelled_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    cancelled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    cancel_reason: Mapped[str | None] = mapped_column(Text)
    closed_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

File: frontend/src/api/replenishment.ts
import axios from 'axios'

import http from './http'

export const REPLENISHMENT_REQUEST_TIMEOUT_MS = 600_000

export interface ReplenishmentRun {
  id: number; run_no: string; calculation_date: string; status: string; default_algorithm: string
  shipment_batch_id: number; inventory_batch_id: number; pipe_wip_batch_id: number
  fitting_wip_batch_id: number; regular_product_batch_id: number; weekly_plan_batch_id: number | null
  total_products: number; suggestion_count: number; positive_suggestion_count: number
  pending_review_count: number; blocking_issue_count: number; warning_issue_count: number
  warning_count: number; approved_count: number; converted_count: number; created_by: number
  source_date_summary: Record<string, string | null>; created_at: string
}

export interface ReplenishmentSuggestion {
  id: number; run_id: number; product_id: number; product_code: string; product_name: string | null
  specification: string | null; category: string | null; unit: string | null
  algorithm: string; monthly_shipments: Record<string, string>
  calculated_target_qty: string; target_stock_qty: string; on_hand_qty: string
  expected_inbound_qty: string; expected_outbound_qty: string; available_qty: string
  pipe_wip_raw_qty: string; pipe_wip_effective_qty: string
  fitting_wip_raw_qty: string; fitting_wip_effective_qty: string; scheduled_not_started_qty: string
  scheduled_known_qty: string; scheduled_override_qty: string
  system_suggested_qty: string; confirmed_qty: string | null; review_status: string
}

export interface ProductionDemand {
  id: number; demand_no: string; product_code: string; product_name: string | null
  specification: string | null; unit: string | null; source_run_no: string; source_suggestion_id: number
  confirmed_qty: string; active_allocated_qty: string; qualified_completed_qty: string
  remaining_to_schedule_qty: string; remaining_to_complete_qty: string; priority: number
  required_date: string | null; status: string; created_at: string
}

export interface SourceBatch { id: number; batch_no: string; import_type: string; source_date: string | null; imported_rows: number; confirmed_at: string; total_staging_rows?: number; matched_rows?: number; ignored_rows?: number; incomplete_rows?: number; matching_complete?: boolean }
export interface ReplenishmentIssue { id: number; run_id: number | null; suggestion_id: number | null; issue_code: string; severity: string; message: string; status: string; product_id: number | null; details: Record<string, unknown> | null }
export interface ProductOption { id: number; product_code: string; product_name: string | null; specification: string | null }

export async function listReplenishmentSourceBatches(importType?: string) {
  const params: Record<string, unknown> = {}
  if (importType) params.import_type = importType
  return (await http.get<{ items: SourceBatch[] }>('/v1/replenishment/source-batches', { params })).data
}

export async function getReplenishmentRun(runId: number) {
  return (await http.get<ReplenishmentRun & { order_inputs: Record<string, unknown>[]; audit_logs: Record<string, unknown>[] }>(`/v1/replenishment/runs/${runId}`)).data
}

export function describeReplenishmentError(error: unknown): string {
  if (axios.isAxiosError(error) && error.code === 'ECONNABORTED') {
    return '计算请求超时，后端可能仍在执行，请刷新运行状态确认，不要重复提交'
  }
  if (axios.isAxiosError(error)) return error.response?.data?.message ?? '补库处理失败，请查看问题清单'
  return '补库处理失败，请稍后重试'
}

export async function listRuns(params: Record<string, unknown> = {}) {
  return (await http.get<{ items: ReplenishmentRun[]; total: number }>('/v1/replenishment/runs', { params })).data
}

export async function createRun(payload: Record<string, unknown>) {
  return (await http.post<ReplenishmentRun>('/v1/replenishment/runs', payload)).data
}

export async function calculateRun(runId: number, override = false, reason?: string) {
  return (await http.post<ReplenishmentRun>(`/v1/replenishment/runs/${runId}/calculate`, {
    override_blocking_checks: override, override_reason: reason,
  }, { timeout: REPLENISHMENT_REQUEST_TIMEOUT_MS })).data
}

export async function listSuggestions(runId: number, params: Record<string, unknown> = {}) {
  return (await http.get<{ items: ReplenishmentSuggestion[]; total: number }>('/v1/replenishment/suggestions', { params: { run_id: runId, ...params } })).data
}

export async function getSuggestion(suggestionId: number) {
  return (await http.get<ReplenishmentSuggestion & { issues: ReplenishmentIssue[] }>(`/v1/replenishment/suggestions/${suggestionId}`)).data
}

export async function listRunIssues(runId: number, params: Record<string, unknown> = {}) {
  return (await http.get<{ items: ReplenishmentIssue[]; total: number }>(`/v1/replenishment/runs/${runId}/issues`, { params })).data
}

export async function resolveRunIssue(runId: number, issueId: number, action: 'RESOLVE' | 'IGNORE', reason: string) {
  return (await http.post(`/v1/replenishment/runs/${runId}/issues/${issueId}`, { action, reason })).data
}

export async function updateScheduledOverride(suggestionId: number, qty: string, reason: string) {
  return (await http.put<ReplenishmentSuggestion>(
    `/v1/replenishment/suggestions/${suggestionId}/scheduled-override`,
    { scheduled_override_qty: qty, reason },
  )).data
}

export async function searchReplenishmentProducts(keyword: string) {
  return (await http.get<{ items: ProductOption[] }>('/v1/replenishment/products', { params: { keyword } })).data
}

export async function reviewSuggestion(suggestionId: number, action: string, confirmedQty: string | undefined, reason: string) {
  return (await http.patch<ReplenishmentSuggestion>(`/v1/replenishment/suggestions/${suggestionId}`, { action, confirmed_qty: confirmedQty, reason })).data
}

export async function bulkReviewSuggestions(suggestionIds: number[], action: string, reason: string) {
  return (await http.post('/v1/replenishment/suggestions/bulk-review', { suggestion_ids: suggestionIds, action, reason })).data
}

export async function convertSuggestions(runId: number, suggestionIds: number[], reason: string) {
  return (await http.post(`/v1/replenishment/runs/${runId}/convert`, { suggestion_ids: suggestionIds, reason }, { timeout: REPLENISHMENT_REQUEST_TIMEOUT_MS })).data
}

export async function approveRun(runId: number, reason: string, allowNoReplenishment = false) {
  return (await http.post<ReplenishmentRun>(`/v1/replenishment/runs/${runId}/approve`, { reason, allow_no_replenishment: allowNoReplenishment })).data
}

export async function listProductionDemands(params: Record<string, unknown> = {}) {
  return (await http.get<{ items: ProductionDemand[]; total: number }>('/v1/production-demands', { params })).data
}

export async function cancelProductionDemand(demandId: number, reason: string) {
  return (await http.post<ProductionDemand>(`/v1/production-demands/${demandId}/cancel`, { reason })).data
}
